### PR TITLE
COMMERCE-481

### PR DESCRIPTION
**Pull request by ZoltanTakacs**
https://issues.liferay.com/browse/COMMERCE-481

### CODE REVIEW

**Comment by ZoltanTakacs** [comment]
ci:test

**Comment by ZoltanTakacs** [comment]
ci:stop

File: modules/etl/talend/talend-definition/src/main/java/com/liferay/talend/runtime/LiferaySourceOrSinkRuntime.java [hunk 1]
```diff
@@ -32,16 +32,30 @@
...
 public interface LiferaySourceOrSinkRuntime extends SourceOrSink {
 
+	public String getActualWebSiteName(String webSiteURL) throws IOException;
+
 	public List<NamedThing> getAvailableWebSites() throws IOException;
 
 	public Schema getExpectedFormSchema(Operation operation) throws IOException;
 
+	/**
+	 * @deprecated In favor of determining the schema based on the resource type
```
Comment (Ithildir) [inline modules/etl/talend/talend-definition/src/main/java/com/liferay/talend/runtime/LiferaySourceOrSinkRuntime.java]: @ZoltanTakacs, why are we deprecating this, if these modules have never been published? Shouldn't we just remove it?

File: modules/etl/talend/talend-definition/src/main/java/com/liferay/talend/utils/URIUtils.java [hunk 2]
```diff
@@ -64,6 +64,15 @@ public static URI getURI(String url) {
 		}
 	}
 
+	public static URI setPaginationLimitOnURL(String resourceURL, int limit) {
+		URI decoratedURI = updateWithQueryParameters(
```
Comment (Ithildir) [inline modules/etl/talend/talend-definition/src/main/java/com/liferay/talend/utils/URIUtils.java]: Return it directly, no need to create a variable.

File: modules/etl/talend/talend-runtime/src/main/java/com/liferay/talend/avro/ResourceCollectionSchemaInferrer.java [hunk 3]
```diff
@@ -96,4 +98,59 @@ public static Schema inferSchemaByResourceFields(
...
+		List<Property> supportedProperties =
+			resourceSupportedClass.getSupportedProperties();
+
+		int size = supportedProperties.size() + 1;
+
+		List<Field> schemaFields = new ArrayList<>(size);
+
+		// Already used names for the fields
+
+		Set<String> filedNames = new HashSet<>();
```
Comment (Ithildir) [inline modules/etl/talend/talend-runtime/src/main/java/com/liferay/talend/avro/ResourceCollectionSchemaInferrer.java]: Typo?

File: modules/etl/talend/talend-runtime/src/main/java/com/liferay/talend/avro/ResourceCollectionSchemaInferrer.java [hunk 4]
```diff
@@ -96,4 +98,59 @@ public static Schema inferSchemaByResourceFields(
 		return schema;
 	}
 
+	public static Schema inferSchemaByResourceType(
+		ApioApiDocumentation.SupportedClass resourceSupportedClass) {
+
+		List<Property> supportedProperties =
+			resourceSupportedClass.getSupportedProperties();
+
+		int size = supportedProperties.size() + 1;
```
Comment (Ithildir) [inline modules/etl/talend/talend-runtime/src/main/java/com/liferay/talend/avro/ResourceCollectionSchemaInferrer.java]: I don't think you're reusing `size` anywhere else, right? Then you can set this directly in the constructor.

File: modules/etl/talend/talend-runtime/src/main/java/com/liferay/talend/avro/ResourceNodeConverter.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+	@Override
+	@SuppressWarnings("unchecked")
+	public IndexedRecord convertToAvro(JsonNode resource) {
+		IndexedRecord record = new GenericData.Record(getSchema());
+
+		Iterator<Map.Entry<String, JsonNode>> jsonFields = resource.fields();
+
+		// Already used names for the fields
+
+		Set<String> normalizedJsonFiledNames = new HashSet<>();
```
Comment (Ithildir) [inline modules/etl/talend/talend-runtime/src/main/java/com/liferay/talend/avro/ResourceNodeConverter.java]: Another typo?

File: modules/etl/talend/talend-runtime/src/main/java/com/liferay/talend/avro/ResourceNodeConverter.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+			}
+
+			if (pos >= 0) {
+				JsonNode resourceJsonNode = field.getValue();
+
+				Object value = avroConverters[pos].convertToAvro(
+					resourceJsonNode.asText());
+
+				record.put(pos, value);
+				pos = -1;
```
Comment (Ithildir) [inline modules/etl/talend/talend-runtime/src/main/java/com/liferay/talend/avro/ResourceNodeConverter.java]: Leave an empty line here.

File: modules/etl/talend/talend-runtime/src/main/java/com/liferay/talend/avro/ResourceNodeConverter.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+			}
+			else {
+				if ("_type".equals(fieldName) || "_context".equals(fieldName)) {
```
Comment (Ithildir) [inline modules/etl/talend/talend-runtime/src/main/java/com/liferay/talend/avro/ResourceNodeConverter.java]: All these nested condition blocks look a little weird, wouldn't it better to just have a single `else if` with:
```java
else if (_log.isDebugEnabled() && !fieldName.equals("_type") && !fieldName.equals("_context")) {
```

File: modules/etl/talend/talend-runtime/src/test/resources/com/liferay/talend/runtime/apio/jsonld/SampleApiDocumentation.json [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+		{
+			"returns": {
+				"@id": "hydra:returns",
+				"@type": "@id"
+			}
+		}
+	],
+	"@id": "/doc",
+	"@type": "ApiDocumentation",
+	"description": "This API allows developers try a Hypermedia API without creating one",
```
Comment (Ithildir) [inline modules/etl/talend/talend-runtime/src/test/resources/com/liferay/talend/runtime/apio/jsonld/SampleApiDocumentation.json]: Isn't it `allows developers TO try a...`"? Not sure 😅

**Comment by ZoltanTakacs** [comment]
Thank you very much @Ithildir for reviewing it! I really appreciate it! 😊 I'll resend it with the requested changes.

**Comment by ZoltanTakacs** [comment]
Resent on https://github.com/Ithildir/liferay-portal/pull/1057

**Pull request closed without merging**